grants_tagger/models.py

```diff
@@ -0,0 +1,604 @@
+    >> get_params_for_component(params, "vec")
+    {"min_df": 1}
+    """
+    component_params = {}
+    for k, v in params.items():
+        if k.startswith(component):
+            _, component_arg = k.split(f"{component}__")
+            component_params[component_arg] = v
+    return component_params
+
+
+class ScienceEnsemble():
+    def __init__(self, threshold=0.5):
+        self.threshold = threshold
+
+    def fit(self, X, Y):
+        raise NotImplementedError
+
+    def predict(self, X):
+        if self.threshold != 0.5:
+            Y_pred = self.model.predict_proba(X) > self.threshold
+        else:
+            Y_pred = self.model.predict(X)
+        return Y_pred
+
+    def predict_proba(self, X):
+        # TODO: Replace with self.model.predict_proba(X) when implmented
+        Y_probs = np.array([
+            est.predict_proba(X) for est in self.estimators])
+        Y_prob = np.mean(Y_probs, axis=0)
+        return Y_prob
+
+    def save(self):
+        raise NotImplementedError
+
+    def _load(self, model_path):
+        # TODO: Retrieve from predefined locations that save will use
+        if "tfidf_svm" in model_path:
+            with open(model_path, "rb") as f:
+                model = pickle.loads(f.read())
+            return model
+        elif "scibert" in model_path:
+            model = BertClassifier(pretrained="scibert")
+            model.load(model_path)
+            return model
+        else:
+            print(f"Did not recognise model in {model_path} to be one of tfidf-svm or scibert")
+            raise NotImplementedError
+
+    def load(self, model_paths):
+        self.estimators = []
+        for model_path in model_paths.split(","):
+            estimator = self._load(model_path)
+            self.estimators.append(estimator)
+        
+        self.model = WellcomeVotingClassifier(
+            estimators=self.estimators,
+            voting="soft",
+            multilabel=True
+        )
+
+
+class MeshCNN():
+    def __init__(self, threshold=0.5, batch_size=256, shuffle=True,
+            buffer_size=1000, data_cache=None, random_seed=42):
+        """
+        threshold: float, default 0.5. Probability threshold on top of which a tag should be assigned.
+        batch_size: int, default 256. Size of batches used for training and prediction.
+        shuffle: bool, default True. Flag on whether to shuffle data before fit.
+        buffer_size: int, default 1000. Buffer size used for shuffling or transforming data before fit.
+        data_cache: path, default None. Path to use for caching data transformations.
+        random_seed: int, default 42. Random seed that controls reproducibility.
+        """
+        self.threshold = threshold
+        self.batch_size = batch_size
+        self.shuffle = shuffle
+        self.buffer_size = buffer_size
+        self.data_cache = data_cache
+        self.random_seed = random_seed
+
+    def _yield_data(self, X, vectorizer, Y=None):
```

Comment:
I have the feeling that there is an opportunity to move this to WellcomeML, as we use generators there too... Let's keep it in mind.

Reply:
yes 100% basically WellcomeML currently expects tensorflow dataset instead of generators but there is no reason why we should not support both